charmbits/httpservice/server.go

```diff
@@ -0,0 +1,439 @@
+	// set by the setFieldValue argument to RegisterRelation
+	// calls.
+	relationState reflect.Value
+	httpListener  *handlerListener
+	httpsListener *handlerListener
+}
+
+// ServerState holds the state of a server - all the
+// parameters that might affect it.
+//
+// It is an implementation detail and only exposed
+// because net/rpc requires it.
+type ServerState struct {
+	HTTPPort  int
+	HTTPSPort int
+	CertPEM   string
+
+	// Arg holds the value passed to Service.Start, marshaled as JSON.
+	ArgData []byte
+
+	// RelationValues holds an entry for each relation
+	// defined in the second argument of the handler function
+	// passed to Register.
+	RelationValues map[string][]byte
+}
+
+func startServer(ctxt *service.Context, args []string, h *handlerInfo) (hook.Command, error) {
+	c, err := startServer0(ctxt, args, h)
+	if err != nil {
+		return c, err
+	}
+	return c, err
+}
+
+// start starts the actual HTTP server running. It is called
+// in service context, not hook context.
+func startServer0(ctxt *service.Context, args []string, h *handlerInfo) (hook.Command, error) {
+	if len(args) != 1 {
+		return nil, fmt.Errorf("need exactly one argument, got %q", args)
+	}
+	srv := &server{
+		handlerInfo: h,
+		stateDir:    args[0],
+	}
+	state, err := srv.loadState()
+	if err != nil {
+		return nil, fmt.Errorf("cannot load state: %v", err)
+	}
+	var fb Feedback
+	srv.set(state, &fb)
+	fb.show()
+	// Fool net/rpc into thinking the type is exported, because
+	// we don't want to export the srv type from this package.
+	type Srv struct {
+		*server
+	}
+	rpcCmd, err := ctxt.ServeLocalRPC(Srv{srv})
+	if err != nil {
+		return nil, errgo.Notef(err, "serve local RPC failed")
+	}
+	srv.tomb.Go(func() error {
+		srv.tomb.Kill(rpcCmd.Wait())
+		return nil
+	})
+	srv.tomb.Go(func() error {
+		<-srv.tomb.Dying()
+		// The command has been killed. Stop the RPC listener
+		// and close any current listeners to cause the handlers
+		// to terminate.
+		rpcCmd.Kill()
+		err := rpcCmd.Wait()
+		if err != nil {
+			err = errgo.Notef(err, "local RPC server")
+		}
+
+		srv.mu.Lock()
+		defer srv.mu.Unlock()
+		srv.closeResources()
+		return err
+	})
+	return srv, nil
+}
+
+// Kill implements gocharm.Command.Kill.
+func (srv *server) Kill() {
+	srv.tomb.Kill(nil)
+}
+
+// Wait implements gocharm.Command.Wait.
+func (srv *server) Wait() error {
+	return srv.tomb.Wait()
+}
+
+// Feedback is an implementation detail, exposed only because net/rpc requires it.
+type Feedback struct {
+	Warnings []string
+}
+
+// show shows all the warnings.
+func (fb *Feedback) show() {
+	for _, w := range fb.Warnings {
+		fmt.Fprintf(os.Stderr, "warning: %v\n", w)
+	}
+}
+
+// addError adds the given error to the list of warnings
+// that will be reported as the result of an RPC call.
+func (fb *Feedback) addError(err error) {
+	fb.Warnings = append(fb.Warnings, err.Error())
+}
+
+// Changed implements an RPC server method that is called
```

Comment:
Done. Also realised that we weren't actually using the values provided in the feedback, so done that too.